### DISCUSSION TITLE
Layout algorithms to use

### DISCUSSION THREAD

**Discussion by chandrasekhar1996** [xyflow/xyflow#1786] [Q&A]
Hey,
In the layout examples provided, two libraries `dagre` and `elkjs` are mentioned. I am ideally looking for a **MIT license** layout library to use with my react application. `dagre` seems to be the widely popular option but it has been deprecated/not actively maintained from this year. 

I would like to ask what is the best alternative to `dagre` for graph layout. Any help is highly appreciated.

**Comment by moklick** [comment]
We made good experience with [cytoscape.js](https://js.cytoscape.org/ ). I haven't tried it with React Flow but [WebCola](https://github.com/tgdwyer/WebCola) could be interesting too. Both libraries have MIT licenses.

**Reply by chandrasekhar1996** [reply]
@moklick sorry, a little confused here. According to my understand, Cytoscape.js is a full fledged graph visualization library similar to react-flow which also provides layout extensions. Are you suggesting to use that library to replace react-flow?

**Reply by moklick** [reply]
It is.. but you can also just its layout algorithms if you don't care too much about bundle size ;) You can also run cytoscape in [headless mode](https://js.cytoscape.org/#init-opts/headless).

**Reply by chandrasekhar1996** [reply]
@moklick I had a doubt regarding how layouting works with react-flow or in general.
When I tried to print the width and height of my nodes (node.__rf.width/height), it is initially null. I am assuming at this point the graph hasn't been rendered yet, hence the value.
After 1-2 iterations, I see the proper values of width and height being printed out, does the layout algorithm supposed to act now after the graph is rendered? If yes, does that mean the graph is rendered again after getting the values of x and y?

I am having this question primarily since if I plan to run cytoscape in headless mode, I need to provide the width and height of my custom node.

I am a little new to such visualization libraries and would highly appreciate your help.

**Reply by blqke** [reply]
Had the same issue @chandrasekhar1996

One solution could be to have a predictable formula to compute your nodes size, but when you have paddings, dynamic content etc in your nodes, it is quickly becoming a pain

So I am now getting the width and height, then using a layout algorithm

(I am using https://klortho.github.io/d3-flextree/ , but this should be applicable to  dagre elkjs cytoscapes..)
